Systemd service for the script is not created and the unit enters into error state

### Bug Description

The charm is currently not running [_create_service_if_not_existing](https://github.com/canonical/script-exporter-operator/blob/main/src/charm.py#L159), so the execution fails:


Manually adding the systemd service and setting the right path for the configuration `--config.file=/etc/script-exporter-config.yaml`, reload the daemon and start the service fix the issue.

Would be good to have a minimal functional test to guarantee that this kind of issue doesn't happen in the future.

```
Model   Controller  Cloud/Region         Version  SLA          Timestamp
ubuntu  lxd         localhost/localhost  3.5.4    unsupported  17:44:44-03:00

App              Version  Status   Scale  Charm            Channel        Rev  Exposed  Message
grafana-agent             blocked      1  grafana-agent    latest/stable  332  no       Missing ['grafana-cloud-config']|['grafana-dashboards-provider']|['logging-consumer']|['send-remote-write'] for cos-a...
script-exporter           error        1  script-exporter  latest/edge      1  no       hook failed: "config-changed"
ubuntu           22.04    active       1  ubuntu           latest/stable   25  no       

Unit                  Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*             active    idle   0        10.1.156.12            
  grafana-agent/0*    blocked   idle            10.1.156.12            Missing ['grafana-cloud-config']|['grafana-dashboards-provider']|['logging-consumer']|['send-remote-write'] for cos-a...
  script-exporter/0*  error     idle            10.1.156.12            hook failed: "config-changed"
```

### To Reproduce

1. Deploy charm ubuntu
2. Deploy charm grafana-agent
3. Deploy charm script-exporter-operator
4. Relate ubuntu with grafana-agent
5. Relate grafana-agent with script-exporter
6. Relate ubuntu with script-exporter
7. Use the same configuration explained in README

### Environment

Juju 3 using lxd.

### Relevant log output

```shell
unit-script-exporter-0: 17:38:35 INFO juju.worker.uniter awaiting error resolution for "config-changed" hook
unit-script-exporter-0: 17:38:35 ERROR unit.script-exporter/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/lib/charms/operator_libs_linux/v1/systemd.py", line 90, in _systemctl
    proc = subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['systemctl', 'restart', 'script-exporter.service']' returned non-zero exit status 5.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/./src/charm.py", line 299, in <module>
    ops.main(ScriptExporterCharm)  # type: ignore
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/__init__.py", line 343, in __call__
    return _main.main(charm_class=charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/_main.py", line 543, in main
    manager.run()
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/_main.py", line 529, in run
    self._emit()
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/_main.py", line 518, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name, self._juju_context)
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/_main.py", line 134, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/framework.py", line 347, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/framework.py", line 857, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/venv/ops/framework.py", line 947, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/./src/charm.py", line 100, in on_config_changed
    service_restart("script-exporter.service")
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/lib/charms/operator_libs_linux/v1/systemd.py", line 177, in service_restart
    return _systemctl("restart", *args, check=True) == 0
  File "/var/lib/juju/agents/unit-script-exporter-0/charm/lib/charms/operator_libs_linux/v1/systemd.py", line 104, in _systemctl
    raise SystemdError(
charms.operator_libs_linux.v1.systemd.SystemdError: Command ['systemctl', 'restart', 'script-exporter.service'] failed with returncode 5. systemctl output:
Failed to restart script-exporter.service: Unit script-exporter.service not found.

unit-script-exporter-0: 17:38:35 ERROR juju.worker.uniter.operation hook "config-changed" (via hook dispatching script: dispatch) failed: exit status 1
```

### Additional context

This was the file `/etc/systemd/system/script-exporter.service` that I created to fix the issue:

```
[Unit]
Description=Prometheus Script exporter
Wants=network-online.target
After=network-online.target

[Service]
LimitNPROC=infinity
LimitNOFILE=infinity
ExecStart=/usr/local/bin/script_exporter --config.file=/etc/script-exporter-config.yaml
Restart=always

[Install]
WantedBy=multi-user.target
```

I installed rev 4 of the charm and I see
```console
2025-06-13 10:55:18 INFO juju unit_agent.go:288 Starting unit workers for "script-exporter/2"
2025-06-13 10:55:18 INFO juju.worker.apicaller connect.go:163 [b95833] "unit-script-exporter-2" successfully connected to "10.150.11.16:17070"
2025-06-13 10:55:18 INFO juju.worker.apicaller connect.go:260 [b95833] password changed for "unit-script-exporter-2"
2025-06-13 10:55:18 INFO juju.worker.apicaller connect.go:163 [b95833] "unit-script-exporter-2" successfully connected to "10.150.13.16:17070"
2025-06-13 10:55:18 INFO juju.worker.upgrader upgrader.go:141 no waiter, upgrader is done
2025-06-13 10:55:18 INFO juju.worker.migrationminion worker.go:147 migration phase is now: NONE
2025-06-13 10:55:18 INFO juju.worker.logger logger.go:120 logger worker started
2025-06-13 10:55:19 ERROR juju.worker.meterstatus runner.go:93 error running "meter-status-changed": charm missing from disk
2025-06-13 10:55:19 INFO juju.worker.uniter uniter.go:363 unit "script-exporter/2" started
2025-06-13 10:55:19 INFO juju.worker.uniter uniter.go:689 resuming charm install
2025-06-13 10:55:19 INFO juju.worker.uniter.charm bundles.go:82 downloading ch:amd64/script-exporter-4 from API server
2025-06-13 10:55:20 INFO juju.worker.uniter uniter.go:389 hooks are retried true
2025-06-13 10:55:21 INFO juju.worker.uniter.storage resolver.go:126 initial storage attachments ready
2025-06-13 10:55:21 INFO juju.worker.uniter resolver.go:185 found queued "install" hook
2025-06-13 10:55:21 INFO unit.script-exporter/2.juju-log server.go:325 Running legacy hooks/install.
2025-06-13 10:55:24 INFO unit.script-exporter/2.juju-log server.go:325 Script Exporter binary file has been obtained from an attached resource.
2025-06-13 10:55:24 INFO juju.worker.uniter.operation runhook.go:186 ran "install" hook (via hook dispatching script: dispatch)
2025-06-13 10:55:25 INFO juju.worker.uniter.operation runhook.go:186 ran "juju-info-relation-created" hook (via hook dispatching script: dispatch)
2025-06-13 10:55:25 INFO juju.worker.uniter.operation runhook.go:186 ran "leader-settings-changed" hook (via hook dispatching script: dispatch)
2025-06-13 10:55:25 ERROR unit.script-exporter/2.juju-log server.go:325 Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/lib/charms/operator_libs_linux/v1/systemd.py", line 90, in _systemctl
    proc = subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['systemctl', 'restart', 'script-exporter.service']' returned non-zero exit status 5.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/./src/charm.py", line 300, in <module>
    ops.main(ScriptExporterCharm)  # type: ignore
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/__init__.py", line 347, in __call__
    return _main.main(charm_class=charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/_main.py", line 565, in main
    manager.run()
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/_main.py", line 550, in run
    self._emit()
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/_main.py", line 502, in _emit
    self._emit_charm_event(self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/_main.py", line 537, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/framework.py", line 348, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/framework.py", line 907, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/venv/ops/framework.py", line 997, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/./src/charm.py", line 102, in on_config_changed
    service_restart("script-exporter.service")
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/lib/charms/operator_libs_linux/v1/systemd.py", line 177, in service_restart
    return _systemctl("restart", *args, check=True) == 0
  File "/var/lib/juju/agents/unit-script-exporter-2/charm/lib/charms/operator_libs_linux/v1/systemd.py", line 104, in _systemctl
    raise SystemdError(
charms.operator_libs_linux.v1.systemd.SystemdError: Command ['systemctl', 'restart', 'script-exporter.service'] failed with returncode 5. systemctl output:
Failed to restart script-exporter.service: Unit script-exporter.service not found.

2025-06-13 10:55:26 ERROR juju.worker.uniter.operation runhook.go:180 hook "config-changed" (via hook dispatching script: dispatch) failed: exit status 1
2025-06-13 10:55:26 INFO juju.worker.uniter resolver.go:180 awaiting error resolution for "config-changed" hook
2025-06-13 10:55:28 INFO juju.worker.uniter resolver.go:180 awaiting error resolution for "config-changed" hook

```

so I had to install it manually

--- a/charmcraft.yaml
+++ b/charmcraft.yaml
@@ -30,6 +30,14 @@ parts:
       craftctl default
       git describe --always > $CRAFT_PART_INSTALL/version
 
+  script-exporter-binary:
+    plugin: dump
+    source: https://github.com/ricoberger/script_exporter/releases/download/v2.15.1/script_exporter-linux-${CRAFT_ARCH_BUILD_FOR}
+    source-type: file
+    permissions:
+      - path: script_exporter-linux-${CRAFT_ARCH_BUILD_FOR}
+        mode: "755"
+
 resources:
   script-exporter-binary:
     type: file

--- a/charmcraft.yaml
+++ b/charmcraft.yaml
@@ -17,8 +17,11 @@ subordinate: true
 
 platforms:
   ubuntu@20.04:amd64:
+  ubuntu@20.04:arm64:
   ubuntu@22.04:amd64:
+  ubuntu@22.04:arm64:
   ubuntu@24.04:amd64:
+  ubuntu@24.04:arm64:
 
 parts:
   charm:

--- a/src/charm.py
+++ b/src/charm.py
@@ -8,15 +8,13 @@ import base64
 import io
 import logging
 import os
+import platform
 import shutil
 import tarfile
 import textwrap
-from hashlib import sha256
 from lzma import LZMAError, decompress
 from pathlib import Path
 from typing import List
-from urllib import request
-from urllib.error import HTTPError
 
 import ops
 import yaml
@@ -31,14 +29,16 @@ from charms.operator_libs_linux.v1.systemd import (
 )
 from ops import ActiveStatus, BlockedStatus, StatusBase
 from ops.model import ModelError
-from ops.pebble import APIError
 
 logger = logging.getLogger(__name__)
 
 EXPORTER_PORT = 9469
 
-EXPORTER_BINARY_URL = "https://github.com/ricoberger/script_exporter/releases/download/v2.15.1/script_exporter-linux-amd64"
-EXPORTER_BINARY_SHA = "e7962a9863c015f721e3cec9af24c85e6b93be79ff992230d9d12029c89f456f"
+
+def _arch():
+    arch = platform.machine()
+    arch = "amd64" if arch == "x86_64" else arch
+    return arch
 
 
 class ScriptExporterCharm(ops.CharmBase):
@@ -52,6 +52,7 @@ class ScriptExporterCharm(ops.CharmBase):
         self._single_script_path = LocalPath("/etc/script-exporter-script")
         self._config_path = LocalPath(f"{self._script_exporter_dir}/config.yaml")
         self._binary_path = LocalPath("/usr/local/bin/script_exporter")
+        self._packaged_binary_path = Path(self.charm_dir) / f"script_exporter-linux-{_arch()}"
         self._binary_resource_name = "script-exporter-binary"
         self._script_daemon_service = Path("/etc/systemd/system/script-exporter.service")
 
@@ -114,12 +115,15 @@ class ScriptExporterCharm(ops.CharmBase):
 
     def _ensure_binary(self) -> None:
         # Make sure the exporter binary is present with a systemd service
-        try:
-            self._obtain_exporter(exporter_url=EXPORTER_BINARY_URL, binary_sha=EXPORTER_BINARY_SHA)
-        except HTTPError as e:
-            msg = f"Script Exporter binary couldn't be downloaded - {str(e)}"
-            logger.error(msg)
-            raise
+        if self._push_exporter_if_attached():
+            return
+
+        shutil.copy(self._packaged_binary_path, self._binary_path)
+        os.chmod(self._binary_path, 0o755)
+        logger.info(
+            "Script Exporter binary installed from packaged files: %s", self._binary_path
+        )
+
 
     def _ensure_scripts_dir(self) -> None:
         # Create the scripts directory if it doesn't exist
@@ -277,19 +281,6 @@ class ScriptExporterCharm(ops.CharmBase):
         # so it will survive reboots
         service_resume("script-exporter.service")
 
-    def _obtain_exporter(self, exporter_url: str, binary_sha: str) -> None:
-        """Obtain Script Exporter binary from an attached resource or download it.
-
-        Args:
-            exporter_url: url of script exporter binary
-            binary_sha: sha of the script exporter binary
-        """
-        # If not coming from resource and not existing already, download the exporter
-        if not self._push_exporter_if_attached() and self._script_exporter_must_be_downloaded(
-            binary_sha
-        ):
-            self._download_exporter_binary(exporter_url)
-
     def _push_exporter_if_attached(self) -> bool:
         """Check whether Script Exporter binary is attached to the charm or not.
 
@@ -311,75 +302,6 @@ class ScriptExporterCharm(ops.CharmBase):
             os.chmod(self._binary_path, 0o755)
             return True
         return False
-
-    def _script_exporter_must_be_downloaded(self, binary_sha: str) -> bool:
-        """Check whether script exporter binary must be downloaded or not.
-
-        Args:
-            binary_sha: string sha of the script exporter binary
-
-        Returns:
-            a boolean representing whether Script Exporter should be downloaded
-        """
-        if not self._is_exporter_binary_in_charm(self._binary_path):
-            return True
-
-        if not self._sha256sums_matches(self._binary_path, binary_sha):
-            return True
-
-        logger.debug("Script Exporter binary file is already in the the charm container.")
-        return False
-
-    def _sha256sums_matches(self, file_path: PathProtocol, sha256sum: str) -> bool:
-        """Check whether a file's sha256sum matches or not with a specific sha256sum.
-
-        Args:
-            file_path: A string representing the files' patch.
-            sha256sum: The sha256sum against which we want to verify.
-
-        Returns:
-            a boolean representing whether a file's sha256sum matches or not with
-            a specific sha256sum.
-        """
-        try:
-            file_bytes = file_path.read_bytes()
-            result = sha256(file_bytes).hexdigest()
-
-            if result != sha256sum:
-                msg = f"File sha256sum mismatch, expected:'{sha256sum}' but got '{result}'"
-                logger.debug(msg)
-                return False
-
-            return True
-        except (APIError, FileNotFoundError):
-            msg = f"File: '{file_path}' could not be opened"
-            logger.error(msg)
-            return False
-
-    def _is_exporter_binary_in_charm(self, binary_path: PathProtocol) -> bool:
-        """Check if Script Exporter binary is already stored locally.
-
-        Args:
-            binary_path: LocalPath of the binary to check
-
-        Returns:
-            a boolean representing whether Script Exporter is present or not.
-        """
-        return True if binary_path.is_file() else False
-
-    def _download_exporter_binary(self, exporter_url: str) -> None:
-        """Download the Script Exporter binary file and move it to its new location.
-
-        Args:
-            exporter_url: url where to get Script Exporter binary from
-        """
-        with request.urlopen(exporter_url) as r:
-            file_bytes = r.read()
-            self._binary_path.write_bytes(file_bytes, mode=0o755)
-            logger.info(
-                "Script Exporter binary file has been downloaded and stored in: %s",
-                self._binary_path,
-            )
 
 
 if __name__ == "__main__":  # pragma: nocover

--- a/src/charm.py
+++ b/src/charm.py
@@ -35,10 +35,11 @@ logger = logging.getLogger(__name__)
 EXPORTER_PORT = 9469
 
 
-def _arch():
-    arch = platform.machine()
-    arch = "amd64" if arch == "x86_64" else arch
-    return arch
+ARCH = platform.machine()
+if platform.machine() == "x86_64":
+    ARCH = "amd64"
+elif platform.machine() in ("aarch64", "arm64", "armv8b", "armv8l"):
+    ARCH = "arm64"
 
 
 class ScriptExporterCharm(ops.CharmBase):
@@ -52,7 +53,6 @@ class ScriptExporterCharm(ops.CharmBase):
         self._single_script_path = LocalPath("/etc/script-exporter-script")
         self._config_path = LocalPath(f"{self._script_exporter_dir}/config.yaml")
         self._binary_path = LocalPath("/usr/local/bin/script_exporter")
-        self._packaged_binary_path = Path(self.charm_dir) / f"script_exporter-linux-{_arch()}"
         self._binary_resource_name = "script-exporter-binary"
         self._script_daemon_service = Path("/etc/systemd/system/script-exporter.service")
 
@@ -118,7 +118,7 @@ class ScriptExporterCharm(ops.CharmBase):
         if self._push_exporter_if_attached():
             return
 
-        shutil.copy(self._packaged_binary_path, self._binary_path)
+        shutil.copy("script_exporter-linux-{}".format(ARCH), self._binary_path)
         os.chmod(self._binary_path, 0o755)
         logger.info(
             "Script Exporter binary installed from packaged files: %s", self._binary_path

--- a/src/charm.py
+++ b/src/charm.py
@@ -41,6 +41,8 @@ if platform.machine() == "x86_64":
 elif platform.machine() in ("aarch64", "armv8b", "armv8l"):
     ARCH = "arm64"
 
+SERVICE_FILENAME = "script-exporter.service"
+
 
 class ScriptExporterCharm(ops.CharmBase):
     """Charm the application."""
@@ -54,7 +56,7 @@ class ScriptExporterCharm(ops.CharmBase):
         self._config_path = LocalPath(f"{self._script_exporter_dir}/config.yaml")
         self._binary_path = LocalPath("/usr/local/bin/script_exporter")
         self._binary_resource_name = "script-exporter-binary"
-        self._script_daemon_service = Path("/etc/systemd/system/script-exporter.service")
+        self._script_daemon_service = Path("/etc/systemd/system/{}".format(SERVICE_FILENAME))
 
         self.cos_agent = COSAgentProvider(
             charm=self,
@@ -79,12 +81,12 @@ class ScriptExporterCharm(ops.CharmBase):
 
     def _on_start(self, _: ops.StartEvent):
         """Handle start event."""
-        service_restart("script-exporter.service")
+        service_restart(SERVICE_FILENAME)
 
     def _on_stop(self, _: ops.StopEvent):
         """Ensure that script exporter is stopped."""
-        if service_running("script-exporter"):
-            service_stop("script-exporter")
+        if service_running(SERVICE_FILENAME):
+            service_stop(SERVICE_FILENAME)
 
         self._remove_file_dir(self._script_exporter_dir)
         self._remove_file_dir(self._binary_path)
@@ -96,7 +98,11 @@ class ScriptExporterCharm(ops.CharmBase):
         self._ensure_scripts_dir()
         self._set_config_file()
         self._set_script_files()
-        service_restart("script-exporter.service")
+
+        if self.model.config["config_file"]:
+            service_restart(SERVICE_FILENAME)
+        elif service_running(SERVICE_FILENAME):
+            service_stop(SERVICE_FILENAME)
 
     def _on_collect_unit_status(self, event: ops.CollectStatusEvent) -> None:
         """Calculate and set the unit status."""
@@ -276,10 +282,10 @@ class ScriptExporterCharm(ops.CharmBase):
         self._script_daemon_service.write_text(systemd_template)
 
         daemon_reload()
-        service_restart("script-exporter.service")
+        service_restart(SERVICE_FILENAME)
         # `enable --now`, but it's the only method which ACTUALLY enables it
         # so it will survive reboots
-        service_resume("script-exporter.service")
+        service_resume(SERVICE_FILENAME)
 
     def _push_exporter_if_attached(self) -> bool:
         """Check whether Script Exporter binary is attached to the charm or not.

--- a/src/charm.py
+++ b/src/charm.py
@@ -81,7 +81,8 @@ class ScriptExporterCharm(ops.CharmBase):
 
     def _on_start(self, _: ops.StartEvent):
         """Handle start event."""
-        service_restart(SERVICE_FILENAME)
+        if not service_running(SERVICE_FILENAME) and self.model.config["config_file"]:
+            service_restart(SERVICE_FILENAME)
 
     def _on_stop(self, _: ops.StopEvent):
         """Ensure that script exporter is stopped."""

--- a/src/charm.py
+++ b/src/charm.py
@@ -38,7 +38,7 @@ EXPORTER_PORT = 9469
 ARCH = platform.machine()
 if platform.machine() == "x86_64":
     ARCH = "amd64"
-elif platform.machine() in ("aarch64", "arm64", "armv8b", "armv8l"):
+elif platform.machine() in ("aarch64", "armv8b", "armv8l"):
     ARCH = "arm64"
 
 

